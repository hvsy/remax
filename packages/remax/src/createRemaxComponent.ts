import React, { forwardRef } from 'react';
import kebabCase from 'lodash.kebabcase';

export default function createRemaxComponent<P = any>(component: string) {
  const Component: React.FC<P> = (props, ref: any) => {
    const { children = [] } = props;
    return React.createElement(
      kebabCase(component).replace('remax-', ''),
      { ...props, ref },
      children
    );
  };
  return forwardRef<{}, React.PropsWithChildren<P>>(Component);
}

export function createCustomRemaxComponent<P = any>(component: string) {
  if (!component.toLowerCase().startsWith('remax')) {
    throw new Error(
      '自定义组件命名请使用 Remax 前缀。如：RemaxCustomComponent'
    );
  }

  return createRemaxComponent<P>(component);
}
