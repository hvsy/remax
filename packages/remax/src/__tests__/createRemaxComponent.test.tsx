import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import { createCustomRemaxComponent } from '../createRemaxComponent';

describe('createCustomRemaxComponent', () => {
  it('create component with remax prefix', () => {
    expect(() => {
      createCustomRemaxComponent<{ foo: string }>('CustomComponent');
    }).toThrowError();

    const RemaxCustomComponent = createCustomRemaxComponent<{ foo: string }>(
      'RemaxCustomComponent'
    );
    const testRenderer = TestRenderer.create(
      <RemaxCustomComponent foo="bar" />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
