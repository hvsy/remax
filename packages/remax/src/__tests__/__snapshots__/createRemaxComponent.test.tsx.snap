// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCustomRemaxComponent create component with remax prefix 1`] = `
<custom-component
  foo="bar"
/>
`;
